File: ios/HazeWalletCore/Sources/HazeWalletCore/SecureStorage.swift
import Foundation
import Security

/// Persists the wallet's keystore (secret) in the iOS Keychain and
/// everything else (the non-secret wallet-store ledger, node/explorer
/// URLs, claimed name, activity log) in UserDefaults - the same
/// secret/non-secret split Android's SecureStorage.kt makes between the
/// Android Keystore and its own SharedPreferences.
public final class SecureStorage {
    public static let defaultNodeUrl = "https://haze-b3l9.onrender.com"
    public static let defaultExplorerUrl = "https://haze-b3l9.onrender.com"

    private let defaults = UserDefaults.standard
    private let keychainAccount = "com.haze.wallet.keystore"

    public init() {}

    // ---------------- keystore (secret, Keychain) ----------------

    public func loadKeystoreBytes() -> Data? {
        var query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: keychainAccount,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        guard status == errSecSuccess, let data = item as? Data else { return nil }
        return data
    }

    public func saveKeystoreBytes(_ bytes: Data) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: keychainAccount,
        ]
        SecItemDelete(query as CFDictionary)
        var addQuery = query
        addQuery[kSecValueData as String] = bytes
        addQuery[kSecAttrAccessible as String] = kSecAttrAccessibleWhenUnlockedThisDeviceOnly
        SecItemAdd(addQuery as CFDictionary, nil)
    }

    private func wipeKeystoreBytes() {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrAccount as String: keychainAccount,
        ]
        SecItemDelete(query as CFDictionary)
    }

    // ---------------- everything else (UserDefaults) ----------------

    public func loadStoreBytes() -> Data? { defaults.data(forKey: "hazeStoreBytes") }
    public func saveStoreBytes(_ bytes: Data) { defaults.set(bytes, forKey: "hazeStoreBytes") }

    public func loadNodeUrl() -> String { defaults.string(forKey: "hazeNodeUrl") ?? Self.defaultNodeUrl }
    public func saveNodeUrl(_ url: String) { defaults.set(url, forKey: "hazeNodeUrl") }

    public func loadExplorerUrl() -> String { defaults.string(forKey: "hazeExplorerUrl") ?? Self.defaultExplorerUrl }
    public func saveExplorerUrl(_ url: String) { defaults.set(url, forKey: "hazeExplorerUrl") }

    public func loadClaimedName() -> String? { defaults.string(forKey: "hazeClaimedName") }
    public func saveClaimedName(_ name: String) { defaults.set(name, forKey: "hazeClaimedName") }

    public func loadActivityLogJson() -> String { defaults.string(forKey: "hazeActivityLog") ?? "[]" }
    public func saveActivityLogJson(_ json: String) { defaults.set(json, forKey: "hazeActivityLog") }

    /// Clears everything - keystore, ledger, claimed name, activity - but
    /// deliberately leaves nodeUrl/explorerUrl alone, same as Android's
    /// lockWallet(), so re-onboarding doesn't lose those settings.
    public func wipe() {
        wipeKeystoreBytes()
        defaults.removeObject(forKey: "hazeStoreBytes")
        defaults.removeObject(forKey: "hazeClaimedName")
        defaults.removeObject(forKey: "hazeActivityLog")
    }
}

File: android/app/src/main/java/com/haze/wallet/SecureStorage.kt
package com.haze.wallet

import android.content.Context
import android.content.SharedPreferences
import android.util.Base64
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey

/**
 * Android Keystore-backed storage for the wallet's serialized bytes
 * (keystore_bytes, store_bytes) plus small app state (node URL, activity
 * log, claimed name). Unlike the web wallet - which derives an AES key from
 * a user password, since browser storage has no hardware-backed secret of
 * its own - this relies on EncryptedSharedPreferences' own Keystore-backed
 * master key, which is already gated behind the device's own lock screen.
 * That's a stronger default on Android than a user-chosen password would be,
 * so no separate password prompt is added here.
 */
class SecureStorage(context: Context) {
    private val prefs: SharedPreferences

    init {
        val masterKey = MasterKey.Builder(context)
            .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
            .build()
        prefs = EncryptedSharedPreferences.create(
            context,
            "haze_wallet_secure_prefs",
            masterKey,
            EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
            EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
        )
    }

    fun hasWallet(): Boolean = prefs.contains(KEY_KEYSTORE)

    fun saveKeystoreBytes(bytes: ByteArray) {
        prefs.edit().putString(KEY_KEYSTORE, Base64.encodeToString(bytes, Base64.NO_WRAP)).apply()
    }

    fun loadKeystoreBytes(): ByteArray? =
        prefs.getString(KEY_KEYSTORE, null)?.let { Base64.decode(it, Base64.NO_WRAP) }

    fun saveStoreBytes(bytes: ByteArray) {
        prefs.edit().putString(KEY_STORE, Base64.encodeToString(bytes, Base64.NO_WRAP)).apply()
    }

    fun loadStoreBytes(): ByteArray? =
        prefs.getString(KEY_STORE, null)?.let { Base64.decode(it, Base64.NO_WRAP) }

    fun saveNodeUrl(url: String) {
        prefs.edit().putString(KEY_NODE_URL, url).apply()
    }

    fun loadNodeUrl(): String = prefs.getString(KEY_NODE_URL, null) ?: DEFAULT_NODE_URL

    fun saveExplorerUrl(url: String) {
        prefs.edit().putString(KEY_EXPLORER_URL, url).apply()
    }

    fun loadExplorerUrl(): String = prefs.getString(KEY_EXPLORER_URL, null) ?: DEFAULT_EXPLORER_URL

    fun saveClaimedName(name: String?) {
        prefs.edit().putString(KEY_NAME, name).apply()
    }

    fun loadClaimedName(): String? = prefs.getString(KEY_NAME, null)

    fun saveActivityLogJson(json: String) {
        prefs.edit().putString(KEY_ACTIVITY, json).apply()
    }

    fun loadActivityLogJson(): String = prefs.getString(KEY_ACTIVITY, null) ?: "[]"

    /** Erases everything - used by "Lock wallet" / "Reset wallet". */
    fun wipe() {
        prefs.edit().clear().apply()
    }

    companion object {
        private const val KEY_KEYSTORE = "keystore_bytes"
        private const val KEY_STORE = "store_bytes"
        private const val KEY_NODE_URL = "node_url"
        private const val KEY_EXPLORER_URL = "explorer_url"
        private const val KEY_NAME = "claimed_name"
        private const val KEY_ACTIVITY = "activity_log"
        const val DEFAULT_NODE_URL = "https://haze-b3l9.onrender.com"
        // The node serves its own embedded block explorer at its own root -
        // same live address as DEFAULT_NODE_URL, no separate deployment.
        const val DEFAULT_EXPLORER_URL = "https://haze-b3l9.onrender.com"
    }
}
